From: Jiri Slaby <jslaby@suse.cz>
Date: Fri, 10 Oct 2014 13:20:43 +0200
Subject: kgr: usb-storage, mark kthread safe
Patch-mainline: submitted for review
References: fate#313296 bnc#899908

This needs a conversion to a wait for completion with timeout. And
when we time out, we switch the thread. Ideally, the kthread should be
converted to a workqueue instead, since I see no reason for having a
kthread here.

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 drivers/usb/storage/usb.c |   11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

--- a/drivers/usb/storage/usb.c
+++ b/drivers/usb/storage/usb.c
@@ -303,12 +303,21 @@ static int usb_stor_control_thread(void
 	struct Scsi_Host *host = us_to_host(us);
 
 	for (;;) {
+		long to;
+
 		usb_stor_dbg(us, "*** thread sleeping\n");
-		if (wait_for_completion_interruptible(&us->cmnd_ready))
+		to = wait_for_completion_interruptible_timeout(&us->cmnd_ready,
+				HZ * 3);
+		if (to < 0)
 			break;
 
 		usb_stor_dbg(us, "*** thread awakened\n");
 
+		if (!to) {
+			kgr_task_safe(current);
+			continue;
+		}
+
 		/* lock the device pointers */
 		mutex_lock(&(us->dev_mutex));
 
